cdap-security/src/main/java/co/cask/cdap/security/store/FileSecureStore.java

```
@@ -0,0 +1,463 @@
+
+    // Get the keystore password
+    password = cConf.get(Constants.Security.Store.FILE_PASSWORD).toCharArray();
+
+    keyStore = locateKeystore(path, password);
+    ReadWriteLock lock = new ReentrantReadWriteLock(true);
+    readLock = lock.readLock();
+    writeLock = lock.writeLock();
+  }
+
+  /**
+   * Stores an element in the secure store.
+   * @param name Name of the element to store
+   * @param data The data that needs to be securely stored
+   * @param description User provided description of the entry.
+   *@param properties Metadata associated with the data  @throws IOException
+   */
+  @Override
+  public void put(String name, byte[] data, String description, Map<String, String> properties) throws IOException {
+    SecureStoreMetadata meta = SecureStoreMetadata.of(name, description, properties);
+    SecureStoreData secureStoreData = new SecureStoreData(meta, data);
+    Key key = null;
+    writeLock.lock();
+    try {
+      if (keyStore.containsAlias(name)) {
+        key = keyStore.getKey(name, password);
+      }
+      keyStore.setKeyEntry(name, new KeyStoreEntry(secureStoreData, meta), password, null);
+      // Attempt to persist the store.
+      flush();
+    } catch (KeyStoreException | NoSuchAlgorithmException | UnrecoverableKeyException e) {
+      // We failed to store the key in the key store. Throw an IOException.
+      throw new IOException("Failed to store the key. ", e);
+    } catch (IOException ioe) {
+      // The key was stored in the keystore successfully but we failed to flush it to the file system.
+      // Remove the key from the keystore and throw an IOException.
+      try {
+        if (key == null) {
+          deleteFromStore(name, password);
+        } else {
+          keyStore.setKeyEntry(name, key, password, null);
+        }
+      } catch (KeyStoreException | UnrecoverableKeyException | NoSuchAlgorithmException e) {
+        LOG.error("Failed to recover the store after a failed flush.");
```

The exception will not make it clear to the caller what the underlying error was. We are throwing IOException with the keystore exception added as suppressed so the actual error will be hidden from the client.